### PR TITLE
Improve testing

### PR DESCRIPTION
**Pull request by matthewmturner**

### CODE REVIEW

**Comment by seddonm1** [comment]
@matthewmturner i had a quick look. I think you should have some expected error message like 
```
#[should_panic(expected = "Could not parse metadata: bad data")]
```

**Comment by matthewmturner** [comment]
> @matthewmturner i had a quick look. I think you should have some expected error message like 
> 
> ```
> 
> #[should_panic(expected = "Could not parse metadata: bad data")]
> 
> ```

I was playing with that but it's a super long error message that was hard to match on. Maybe I have to extract one of the errors.

**Comment by seddonm1** [comment]
The expected is actually an expected.contains message so you don't need the whole thing but we need to make sure we are catching the correct error.

**Comment by matthewmturner** [comment]
@seddonm1 got it, thx for tip!

File: src/object_store/aws.rs [hunk 1]
```diff
@@ -537,4 +537,117 @@ mod tests {
...
+            None,
+            None,
+        )
+        .await;
+        let mut files = amazon_s3_file_system
+            .list_file("data/nonexistent_file.txt")
+            .await
+            .unwrap();
+
+        if let Some(file) = files.next().await {
```
Comment (seddonm1) [inline src/object_store/aws.rs:636]: Could you just do something like this below (untested) and remove the panic (and all this other non-used code)?

```
assert_eq!(files.next().await, None)
```

Comment (matthewmturner) [reply src/object_store/aws.rs:636]: Thats much nicer, will do

File: src/object_store/aws.rs [hunk 2]
```diff
@@ -537,4 +537,99 @@ mod tests {
...
+
+            let mut bytes = Vec::new();
+            let size = reader.read_to_end(&mut bytes).unwrap();
+
+            assert_eq!(size as u64, sized_file.size);
+        }
+    }
+
+    #[tokio::test]
+    // #[should_panic(expected = "File is missing")]
```
Comment (seddonm1) [inline src/object_store/aws.rs:611]: remove this comment.

**Comment by seddonm1** [comment]
Looks good. I think we should just add a quick comment above each test case to help future readers understand the intent of the test. I don't think the test name is always clear enough.

**Comment by matthewmturner** [comment]
@seddonm1 added test documentation

File: src/object_store/aws.rs [hunk 3]
```diff
@@ -302,6 +302,7 @@ mod tests {
     const PROVIDER_NAME: &str = "Static";
     const MINIO_ENDPOINT: &str = "http://localhost:9000";
 
+    // Test that `AmazonS#FileSystem` can read files
```
Comment (seddonm1) [inline src/object_store/aws.rs:305]: typo

File: src/object_store/aws.rs [hunk 4]
```diff
@@ -492,6 +496,7 @@ mod tests {
         Ok(())
     }
 
+    // Test that the appropriate error message is generated if a Parquet file with bad data is read
```
Comment (seddonm1) [inline src/object_store/aws.rs:499]: Test that the `AmazonS3FileSystem` allows reading from different buckets

File: src/object_store/aws.rs [hunk 5]
```diff
@@ -537,4 +542,101 @@ mod tests {
 
         table.scan(&None, &[], Some(1024)).await.unwrap();
     }
+
+    // Test that `AmazonS3FileSystem` can be registered as object store on a DataFusion `ExecutionContext`
```
Comment (seddonm1) [inline src/object_store/aws.rs:546]: 👍

**Review by seddonm1 — APPROVED** [review]
nice. once these are done we are good.

**Pull request merged**